### PR TITLE
fix(worker): Flush the stream writer on error (DGRAPH-2499)

### PR DESCRIPTION
**Pull request by jarifibrahim**
The stream write blocks writes to badger and if we don't call `writer.flush` we will never unblock the writes

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6609)

 
 

Docs Preview: [Dgraph Preview](https://dgraph-cb70dc08c8-107659.surge.sh)

### CODE REVIEW

File: worker/snapshot.go [hunk 1]
```diff
@@ -66,13 +66,23 @@ func (n *node) populateSnapshot(snap pb.Snapshot, pl *conn.Pool) (int, error) {
...
 			return 0, err
 		}
 
 		writer = sw
+		defer func() {
+			if flushed {
+				return
+			}
+			// Flush the writer if it has not been flushed so that we can unblock the writes.
+			if err := writer.Flush(); err != nil {
```
Comment (jarifibrahim) [inline worker/snapshot.go:82]: We could add a cancel API on stream writer to fix this in a more elegant way.

**Review by manishrjain — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6609)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @jarifibrahim, @manishrjain, and @vvbalaji-dgraph)

___
*[worker/snapshot.go, line 82 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6609#-MIV9vYlpwupEx5fkO9Z-r1-82:-MIVJkhr9Mm_MfMcaDJy:bssd0a9) ([raw file](https://github.com/dgraph-io/dgraph/blob/01b560aac9969ce8ebdefde5c54d82dfbc211231/worker/snapshot.go#L82)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

We could add a cancel API on stream writer to fix this in a more elegant way.
</blockquote></details>

Add a Cancel API as you suggested.

**Review by manishrjain — APPROVED** [review]
:lgtm: Whenever you fix it, you can merge it.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6609#-:-MIZpBBy0CbFXpthlMnI:b-l49yxm)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @jarifibrahim and @vvbalaji-dgraph)

File: worker/snapshot.go (hunk 1 above)
Comment (jarifibrahim) [reply worker/snapshot.go]: @ajeetdsouza is working on resolving this.

**Comment by jarifibrahim** [comment]
Closing this PR since we don't need it anymore.

**Pull request merged**